### PR TITLE
feat: add IsAutoUnDelegate field to CrossStakeDistributeUndelegatedSynPackage and  prevent too many failed in auto refund

### PR DESCRIPTION
**Pull request by j75689**
## Description

1. feat: add IsAutoUnDelegate field to CrossStakeDistributeUndelegatedSynPackage
2. fix: disable undelegate after SecondSunsetFork and prevent too many failed in auto refund

### CODE REVIEW

File: x/stake/endblock.go [hunk 1]
```diff
@@ -304,15 +306,34 @@ func handleRefundStake(ctx sdk.Context, sideChainPrefix []byte, k keeper.Keeper)
...
+				"amount", delegation.GetShares().String(),
+				"sideChainId", bscSideChainId,
+				"result", fmt.Sprintf("%+v", result),
+			)
+			// this is to prevent too many delegation is in unbounded state
+			// if too many delegation is in unbounded state, it will cause too many iteration in the block
+			failedCount++
+			if failedCount >= maxProcessedRefundFailed {
+				break
+			}
```
Comment (zjubfd) [inline x/stake/endblock.go:322]: should continue here?

File: x/stake/keeper/delegation.go [hunk 2]
```diff
@@ -739,6 +739,20 @@ func (k Keeper) CompleteUnbonding(ctx sdk.Context, delAddr sdk.AccAddress, valAd
...
+	key := GetAutoUnDelegateIndexKey(delAddr, valAddr)
+
+	return store.Has(key)
+}
+
+func (k Keeper) SetAutoUnDelegate(ctx sdk.Context, delAddr sdk.AccAddress, valAddr sdk.ValAddress) {
+	store := ctx.KVStore(k.storeKey)
+
+	key := GetAutoUnDelegateIndexKey(delAddr, valAddr)
+	store.Set(key, []byte{}) // index, store empty bytes
```
Comment (zjubfd) [inline x/stake/keeper/delegation.go:753]: will this work to store nothing?

File: x/stake/endblock.go (hunk 1 above)
Comment (j75689) [reply x/stake/endblock.go:322]: fixed

File: x/stake/keeper/delegation.go [hunk 3]
```diff
@@ -876,10 +890,11 @@ func (k Keeper) crossDistributeUndelegated(ctx sdk.Context, delAddr sdk.AccAddre
...
 	transferPackage := types.CrossStakeDistributeUndelegatedSynPackage{
-		EventType: types.CrossStakeTypeDistributeUndelegated,
-		Amount:    bscTransferAmount,
-		Recipient: recipient,
-		Validator: valAddr,
+		EventType:        types.CrossStakeTypeDistributeUndelegated,
+		Amount:           bscTransferAmount,
+		Recipient:        recipient,
+		Validator:        valAddr,
+		IsAutoUnDelegate: k.IsAutoUnDelegate(ctx, delAddr, valAddr),
```
Comment (zjubfd) [inline x/stake/keeper/delegation.go:897]: I guess this is a hardfork logic?

File: x/stake/keeper/delegation.go (hunk 2 above)
Comment (j75689) [reply x/stake/keeper/delegation.go:753]: image
there is another case using the syntax

**Review by pythonberg1997 — APPROVED** [review]

**Review by unclezoro — APPROVED** [review]

**Pull request merged**